### DISCUSSION TITLE
Exercise: Utilizing map and filter in JavaScript for Lab Simulation Data Manipulation

### DISCUSSION THREAD

**Discussion by akash-coded** [akash-coded/mern#206] [Tasks]
#### Context:
In the virtual lab simulation project, developers will often encounter scenarios where they need to manipulate arrays of data, such as lab results, configuration parameters, or simulation states. JavaScript’s `map` and `filter` functions are powerful tools for such manipulations.

#### Objective:
Understand and apply the `map` and `filter` functions in JavaScript to manipulate lab simulation data effectively.

---

#### Task:
Given an array of pendulum lengths, calculate the period of oscillation for each using the formula: 
\[ T = 2\pi \sqrt{\frac{L}{g}} \]
where:
- \( T \) is the period,
- \( L \) is the length of the pendulum,
- \( g \) is the gravitational acceleration (approx. \( 9.8 \, \text{m/s}^2 \)).

#### Example Data:
```javascript
const pendulumLengths = [1, 1.5, 2, 2.5];  // lengths in meters
```

#### Hints:
- Use `Math.PI` for the value of \(\pi\).
- Utilize the `map` function to apply the formula to each length.

---

#### Task:
Given an array of titration results, filter out the results where the pH is neutral (i.e., pH = 7).

#### Example Data:
```javascript
const titrationResults = [6.5, 7, 7, 8.2, 6.8, 7];
```

#### Hints:
- Use the `filter` function to exclude results where pH equals 7.

---

#### Task:
Given an array of objects representing different lenses, calculate the focal length for each and filter out those with negative focal lengths.

#### Example Data:
```javascript
const lenses = [
  { id: 1, R1: 10, R2: -10, n: 1.5 },
  { id: 2, R1: 15, R2: 15, n: 1.6 },
  // ...
];
```
where:
- \( R1 \) and \( R2 \) are the radii of curvature for the lens surfaces,
- \( n \) is the refractive index.

The lensmaker's equation is given by:
\[ \frac{1}{F} = (n - 1) \left( \frac{1}{R1} - \frac{1}{R2} \right) \]

#### Hints:
- Use `map` to calculate the focal length for each lens.
- Use `filter` to exclude lenses with negative focal lengths.

---

#### Task:
Given arrays of microscope configurations and observed cells, link each cell observation to a configuration, and filter out observations where the cell was not visible.

#### Example Data:
```javascript
const configurations = [
  { id: 1, zoom: 10, lightIntensity: 15 },
  { id: 2, zoom: 20, lightIntensity: 20 },
  // ...
];

const observations = [
  { configId: 1, cellId: 'A', visible: true },
  { configId: 2, cellId: 'B', visible: false },
  // ...
];
```

#### Hints:
- Use `map` to create a new array combining configurations and observations.
- Use `filter` to exclude observations where `visible` is `false`.

---

### Notes:
- Ensure to test each solution in a JavaScript environment (browser console, Node.js, etc.) to validate the results.
- Reflect on how these data manipulations might be applicable in the broader context of the virtual lab simulations.
- Consider how these functions might be used in tandem with user interactions, UI updates, and real-time data adjustments in the simulations.

---

#### Task:
Given an array of experiments, each containing an array of results, calculate the average result for each experiment and filter out any experiments with an average below a certain threshold.

#### Example Data:
```javascript
const experiments = [
  { id: 1, results: [2, 3, 3, 4] },
  { id: 2, results: [5, 6, 5, 6] },
  // ...
];
```

#### Hints:
- Use nested `map` functions to calculate the average for each experiment.
- Use `filter` to exclude experiments below the threshold.

---

#### Task:
Given an array of pendulum simulation data, transform the data into a format suitable for visualization (e.g., plotting length vs. period) and filter out any outliers.

#### Example Data:
```javascript
const pendulumData = [
  { length: 1, period: 2.01 },
  { length: 1.5, period: 2.47 },
  // ...
];
```

#### Hints:
- Use `map` to transform the data into an array of `{x, y}` objects.
- Use `filter` to exclude outliers based on a defined criterion.

---

#### Task:
Given arrays of lab configurations and simulation results, correlate each result to a configuration and filter out results that do not match any configuration.

#### Example Data:
```javascript
const configurations = [
  { id: 1, variableA: 10, variableB: 15 },
  // ...
];

const simulationResults = [
  { configId: 1, result: 25 },
  { configId: 3, result: 30 },
  // ...
];
```

#### Hints:
- Use `map` to correlate results with configurations.
- Use `filter` to exclude results without matching configurations.

---

### Notes:
- **Scenario 5** demonstrates handling nested data structures, which is common in complex applications.
- **Scenario 6** provides a practical example of preparing data for visualization, a key aspect of the lab simulations.
- **Scenario 7** illustrates how to correlate different sets of data, which is crucial for linking configurations and results in the simulations.

These scenarios are designed to be progressively more complex and closer to real-world applications, ensuring that learners gradually apply `map` and `filter` in contexts relevant to the virtual lab project. Always encourage learners to explore variations and additional complexities to further deepen their understanding and skills.

**Comment by UmarAikawa** [comment]
// Scenario 1: Simple Data Manipulation with map
const pendulumLengths = [1, 1.5, 2, 2.5];  // lengths in meters
function calcPeriod(pendLength){
    return (2 * Math.PI * Math.sqrt(pendLength / 9.8)).toFixed(2);
}
const periodsOfOscillation = pendulumLengths.map(ln => calcPeriod(ln));
console.log(periodsOfOscillation);

// Scenario 2: Filtering Data Based on Conditions
const titrationResults = [6.5, 7, 7, 8.2, 6.8, 7];
const neutrals = titrationResults.filter(titaVal => titaVal == 7);
console.log(neutrals);

// Scenario 3: Complex Data Manipulation with map and filter
const lenses = [
    { id: 1, R1: 10, R2: -10, n: 1.5 },
    { id: 2, R1: 15, R2: 12, n: 1.6 },
    { id: 3, R1: 25, R2: 18, n: 1.9 },
    { id: 4, R1: -10, R2: 20, n: 1.7 },
    { id: 5, R1: 16, R2: -20, n: 1.2 }
];
function calcFlength(lens){
    return 1 / ((lens.n - 1) * ((1 / lens.R1) - (1 / lens.R2)));
}
function isNegative(lens){
    return calcFlength(lens) < 0;
}
const focalLenses = lenses.map(f => calcFlength(f));
console.log(focalLenses);
const negativeFLenses = lenses.filter(f => isNegative(f));
console.log(negativeFLenses);

// Scenario 4: Compound Data Manipulation
const configurations = [
    { id: 1, zoom: 10, lightIntensity: 15 },
    { id: 2, zoom: 20, lightIntensity: 20 },
    { id: 3, zoom: 30, lightIntensity: 25 },
    { id: 4, zoom: 40, lightIntensity: 30 },
    { id: 5, zoom: 50, lightIntensity: 35 }
];

const observations = [
    { configId: 1, cellId: 'A', visible: true },
    { configId: 2, cellId: 'B', visible: false },
    { configId: 3, cellId: 'C', visible: true },
    { configId: 4, cellId: 'D', visible: false },
    { configId: 5, cellId: 'E', visible: true } 
];

function getObervation(config){
    return observations.filter(o => o.configId == config.id);   
}

function notVisible(observation){
    return observation.visible === false;
}
const comb = [];
const combined = configurations.map(config => comb[config.id] = getObervation(config));
console.log(comb);

const invisibles = observations.filter(o => notVisible(o) === true);
console.log(invisibles);

// Scenario 5: Nested Data Manipulation with map and filter
const experiments = [
    { id: 1, results: [2, 3, 3, 4] },
    { id: 2, results: [5, 6, 5, 6] },
    { id: 3, results: [2, 3, 4, 5] },
    { id: 4, results: [3, 4, 5, 6] },
    { id: 5, results: [4, 5, 6, 3] }
  ];

  function calcAverage(results) {
    let tot = 0;
    results.forEach(n => { tot += n});
    return tot / results.length;
  }

  const averages = experiments.map(function (e) { 
    return {id: e.id, results: e.results, average: calcAverage(e.results)}
  });
  console.log(averages);
  
  const threshold = 4;
  const belowThreshold = averages.filter(av => av.average < threshold);
  console.log(belowThreshold);  

//Scenario 6: Data Transformation and Visualization Preparation
const pendulumData = [
    { length: 1, period: 2.01 },
    { length: 1.5, period: 2.47 },
    { length: 2, period: 2.98 },
    { length: 2.5, period: 3.54 },
    { length: 3, period: 4.15 },
    { length: 3.5, period: 4.81 },
    { length: 4, period: 5.52 }
];

const visual = pendulumData.map(function (v) {
    return '{'+v.length+', '+v.period+'}';
});

console.log(visual);
const criterion = 3; //period should be greater than or equal to 3
const outliers = pendulumData.filter(p => p.period >= criterion);
console.log(outliers);

// Scenario 7: Configuration and Simulation
const configs = [
    { id: 1, variableA: 10, variableB: 15 },
    { id: 2, variableA: 20, variableB: 25 },
    { id: 3, variableA: 30, variableB: 35 },
    { id: 4, variableA: 40, variableB: 45 },
    { id: 5, variableA: 50, variableB: 55 } 
];
  
  const simulationResults = [
    { configId: 1, result: 25 },
    { configId: 2, result: 30 },
    { configId: 7, result: 55 },
    { configId: 6, result: 50 },
    { configId: 3, result: 35 },
    { configId: 8, result: 60 },
    { configId: 4, result: 40 },
    { configId: 12, result: 80 }, 
    { configId: 5, result: 45 },
  ];

function getSimResult(config){
    return simulationResults.filter(d => d.configId == config.id);
}
const correlated = configs.map(function(c) {
    return {id: c.id, result: getSimResult(c)[0].result};   
})

console.log(correlated);
function getConfig(result){
    return configs.filter(o => o.id == result.configId);   
}

const unmatched = simulationResults.filter(function(sr){
    const resConf = getConfig(sr);
    if (resConf.length == 0){
        return sr;
    }
});

console.log(unmatched);
